### PR TITLE
feat(core): [CYBERSOURCE] Add original authorized amount in router data

### PR DESCRIPTION
**Pull request by deepanshu-iiitu**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Add original authorized amount in router data for recurring mandate payments.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

https://github.com/juspay/hyperswitch/issues/3418

## How did you test it?

Testing can be done by creating a recurring mandate payment in cybersource and using the following discover cards: 6011111111111117, 6011000991300009

Curls:

- Payment Intent Create:
```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: API_KEY_HERE' \
--data-raw '{
    "amount": 0,
    "order_details": null,
    "currency": "USD",
    "confirm": false,
    "capture_method": "automatic",
    "authentication_type": "no_three_ds",
    "customer_id": "cut2",
    "email": "johndoe@gmail.com",
    "description": "Hello this is description",
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        }   
    },
    "metadata": {},
    "setup_future_usage": "off_session"
}'
```

- Payments - Confirm: 
`curl --location 'http://localhost:8080/payments/{payment_id}/confirm' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: API_KEY_HERE' \
--data '{
    "setup_future_usage": "off_session",
    "payment_method": "card",
    "payment_method_type": "credit",
    "payment_type": "setup_mandate",
    "off_session": true,
    "payment_method_data": {
        "card": {
            "card_number": "6011000991300009",
            "card_exp_month": "12",
            "card_exp_year": "30",
            "card_holder_name": "joseph Doe",
            "card_cvc": "123"
        }
    },
    "browser_info": {
        "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36",
        "accept_header": "text\\/html,application\\/xhtml+xml,application\\/xml;q=0.9,image\\/webp,image\\/apng,*\\/*;q=0.8",
        "language": "en-GB",
        "color_depth": 24,
        "screen_height": 1440,
        "screen_width": 2560,
        "time_zone": -330,
        "java_enabled": true,
        "java_script_enabled": true,
        "ip_address": "0.0.0.0"
    },
    "mandate_data": {
        "customer_acceptance": {
            "acceptance_type": "offline",
            "accepted_at": "1963-05-03T04:07:52.723Z",
            "online": {
                "ip_address": "125.0.0.1", 
                "user_agent": "amet irure esse"
            }
        },
        "mandate_type": {
            "single_use": {
                "amount": 68607706,
                "currency": "USD"
            }
        }
    }
}'`

- Payments - Create using Mandates
`curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: API_KEY_HERE' \
--data-raw '{
    "amount": 200,
    "currency": "USD",
    "confirm": true,
    "customer_id": "{customer_id}",
    "email": "johndoe@gmail.com",
    "name": "John Doe",
    "phone": "999999999",
  "capture_method": "automatic",
    "phone_country_code": "+1",
    "description": "Its my first payment request",
    "return_url": "https://google.com",
    "mandate_id": "{mandate_id}",
    "off_session":true,
    "browser_info": {
        "ip_address":"172.0.0.1"
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        }     
    }
}'`

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/core/payments/helpers.rs [hunk 1]
```diff
@@ -475,6 +475,27 @@ pub async fn get_token_for_recurring_mandate(
...
+        .original_payment_id
+        .as_ref()
+        .async_map(|payment_id| async {
+            db.find_payment_intent_by_payment_id_merchant_id(
+                payment_id,
+                &mandate.merchant_id,
+                merchant_account.storage_scheme,
+            )
+            .await
+            .to_not_found_response(errors::ApiErrorResponse::MandateNotFound)
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/helpers.rs:488]:
```suggestion
            .to_not_found_response(errors::ApiErrorResponse::PaymentNotFound)
```

**Review by Narayanbhat166 — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

File: crates/router/src/connector/cybersource/transformers.rs [hunk 2]
```diff
@@ -1296,7 +1331,8 @@ impl ForeignFrom<(CybersourcePaymentStatus, bool)> for enums::AttemptStatus {
     fn foreign_from((status, capture): (CybersourcePaymentStatus, bool)) -> Self {
         match status {
             CybersourcePaymentStatus::Authorized
-            | CybersourcePaymentStatus::AuthorizedPendingReview => {
+            | CybersourcePaymentStatus::AuthorizedPendingReview
+            | CybersourcePaymentStatus::Accepted => {
```
Comment (ArjunKarthik) [inline crates/router/src/connector/cybersource/transformers.rs:1335]: Why are we getting a new status from connector for mandate? Please get it clarified from them

File: crates/router/src/connector/cybersource/transformers.rs [hunk 3]
```diff
@@ -1159,13 +1192,14 @@ impl TryFrom<&CybersourceRouterData<&types::PaymentsIncrementalAuthorizationRout
...
+                    initiator: Some(CybersourcePaymentInitiator {
                         initiator_type: None,
                         credential_stored_on_file: None,
                         stored_credential_used: Some(true),
-                    },
+                    }),
                     merchant_intitiated_transaction: Some(MerchantInitiatedTransaction {
-                        reason: "5".to_owned(),
+                        reason: Some("5".to_owned()),
+                        original_authorized_amount: None,
```
Comment (ArjunKarthik) [inline crates/router/src/connector/cybersource/transformers.rs:1202]: Incase of incremental auth don't we need to pass the original_authorized_amount?

File: crates/router/src/connector/cybersource/transformers.rs [hunk 4]
```diff
@@ -470,35 +473,66 @@ impl From<&CybersourceRouterData<&types::PaymentsCompleteAuthorizeRouterData>>
...
-                        },
+                        }),
                         merchant_intitiated_transaction: None,
                     }),
                 )
+            } else if item.router_data.request.connector_mandate_id().is_some() {
+                let original_amount = item
+                    .router_data
+                    .get_recurring_mandate_payment_data()?
+                    .original_payment_authorized_amount
```
Comment (ArjunKarthik) [inline crates/router/src/connector/cybersource/transformers.rs:506]: Please move this to utils

File: crates/router/src/connector/cybersource/transformers.rs [hunk 5]
```diff
@@ -470,35 +473,66 @@ impl From<&CybersourceRouterData<&types::PaymentsCompleteAuthorizeRouterData>>
...
+                    .ok_or(errors::ConnectorError::MissingRequiredField {
+                        field_name: "original_payment_authorized_amount",
+                    })?;
+                let original_currency = item
+                    .router_data
+                    .get_recurring_mandate_payment_data()?
+                    .original_payment_authorized_currency
```
Comment (ArjunKarthik) [inline crates/router/src/connector/cybersource/transformers.rs:513]: Please move this to utils

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/cybersource/transformers.rs (hunk 2 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/cybersource/transformers.rs:1335]: Query Raised

File: crates/router/src/connector/cybersource/transformers.rs (hunk 3 above)
Comment (SamraatBansal) [reply crates/router/src/connector/cybersource/transformers.rs:1202]: @sai-harsha-vardhan

Comment (sai-harsha-vardhan) [reply crates/router/src/connector/cybersource/transformers.rs:1196]: @ArjunKarthik We don't have to send `originalAuthorizedAmount` for incremental auth

**Review by SamraatBansal — APPROVED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Pull request merged**